File: Go/Grpc/node/src/log/log.ts
import pino, { TransportMultiOptions, TransportTargetOptions } from 'pino';

// 使用ES6模块和单例模式来管理pino实例
let loggerInstance: pino.Logger | null = null;

// 日志初始化
function getLogger(): pino.Logger {
	if (!loggerInstance) {
		// 取得 NODE_ENV 环境变量
		const nodeEnv = process.env.NODE_ENV;
		if (nodeEnv === 'development') {
			// 其他，采用漂亮打印
			loggerInstance = prettyConsole();
		} else if (nodeEnv === 'production') {
			// Docker环境，采用json结构打印
			loggerInstance = jsonConsole();
		} else {
			// 其他，采用漂亮打印
			loggerInstance = prettyConsole();
		}
		console.log("pino日志初始化完成");
	}
	return loggerInstance;
}

// 采用漂亮打印
function prettyConsole(): pino.Logger {
	// 从配置文件读取
	const min_level: string = "info";
	// 配置输出
	let targetOptions: TransportTargetOptions[] = [];
	// 控制台
	targetOptions.push({ target: 'pino-pretty', });
	// 文件
	//targetOptions.push({ target: 'pino/file', options: { destination: filename, mkdir: true } });
	// 配置日志选项
	const transportOption: TransportMultiOptions = {
		targets: targetOptions,
		options: {
			colorize: true,
			singleLine: true,
			translateTime: 'yyyy-mm-dd HH:MM:ss.l',
		}
	};
	// 日志对象
	const loggerInstance = pino({
		level: min_level,
		transport: transportOption
	});
	return loggerInstance;
}

// 采用Json日志
function jsonConsole(): pino.Logger {
	// 从配置文件读取
	const min_level: string = "info";
	// 日志对象
	const loggerInstance = pino({
		level: min_level,
		base: { service_name: 'NodeGrpcService' }, // 添加基本字段
		messageKey: 'message' // 为Key改名
	});
	return loggerInstance;
}

// 公开 getLogger
export { getLogger };

File: Go/Grpc/python/src/log/log.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

import os
import sys
from loguru import logger
import json


# 初始化日志
async def init_logger_async() -> None:

    # 获取 PYTHON_ENVIRONMENT 环境变量的值
    env = os.getenv("PYTHON_ENVIRONMENT", "development")
    if env == "docker":
        # Docker环境，采用json结构打印
        await json_console()
    else:
        # 其他，采用漂亮打印
        await pretty_console()


# 采用漂亮打印
async def pretty_console() -> None:
    # 移除默认的日志处理器
    logger.remove()
    # 添加一个自定义的日志处理器，漂亮打印
    logger.add(
        sys.stdout,
        colorize=True,
        format="<green>{time:YYYY-MM-DD HH:mm:ss.SSS}</green> | {level} | {thread} | <cyan>{name}</cyan>:<cyan>{function}</cyan>:<cyan>{line}</cyan> - {message}",
        level="INFO",
        enqueue=True,
        backtrace=True,
        diagnose=True,
    )


# 采用Json日志
async def json_console() -> None:
    # 移除默认的日志处理器
    logger.remove()
    # 添加一个自定义的日志处理器，输出 JSON 格式的日志
    logger.add(
        sink=sink_serializer,
        level="INFO",
        serialize=True,
        enqueue=True,
        backtrace=True,
        diagnose=True,
    )


# 添加一个自定义的日志处理器，输出带有更多字段的 JSON 格式日志
def sink_serializer(message):
    # 将日志消息转换为字典
    message_dict = message.record
    # 序列化为 JSON 字符串
    simplified = {
        "time": message_dict["time"].timestamp(),
        "level": message_dict["level"].name,
        "message": message_dict["message"],
        "exception": message_dict["exception"],
        "file": message_dict["file"].name,
        "function": message_dict["function"],
        "line": message_dict["line"],
        "thread": message_dict["thread"].id,
        # 添加自定义字段
        "service_name": "PythonGrpcService",
    }
    json_message = json.dumps(simplified, default=str, ensure_ascii=False)
    print(json_message, file=sys.stdout, flush=True)
